### PR TITLE
Improve rerouting logic if all candidates are very different from the initial route

### PR DESCRIPTION
**Pull request by azarovalex**
Fixes https://github.com/mapbox/mapbox-navigation-ios/issues/3597

### CODE REVIEW

File: Sources/TestHelper/DummyRoute.swift [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+public class DummyRoute: Route {
+
+    public init(description: String) {
+        let leg = RouteLeg(steps: [],
+                           name: description,
+                           distance: 0,
+                           expectedTravelTime: 0,
+                           typicalTravelTime: nil,
+                           profileIdentifier: .automobile)
+        super.init(legs: [leg], shape: nil, distance: 0, expectedTravelTime: 0, typicalTravelTime: nil)
```
Comment (azarovalex) [inline Sources/TestHelper/DummyRoute.swift:14]: Little hack, because `description` property cannot be overriden

File: Sources/MapboxCoreNavigation/Router.swift [hunk 2]
```diff
@@ -404,10 +404,18 @@ extension InternalRouter where Self: Router {
...
-            let leftDistance = left.element.description.minimumEditDistance(to: target)
-            let rightDistance = right.element.description.minimumEditDistance(to: target)
-            return leftDistance < rightDistance
-        }?.offset
+        
+        guard let bestCandidate = map({
+            (route: $0, editDistance: $0.description.minimumEditDistance(to: target))
+        }).enumerated().min(by: { $0.element.editDistance < $1.element.editDistance }) else { return nil }
+
+
```
Comment (bamx23) [inline Sources/MapboxCoreNavigation/Router.swift:412]:
```suggestion

```

File: Sources/MapboxCoreNavigation/Router.swift [hunk 3]
```diff
@@ -404,10 +404,18 @@ extension InternalRouter where Self: Router {
...
+        // If the most similar route is still more than 50% different from the original route,
+        // we fallback to the fastest route which index is 0.
+        guard bestCandidate.element.route.description.count > 0 else { return 0 }
+        let similarityScore = Double(bestCandidate.element.editDistance) / Double(bestCandidate.element.route.description.count)
```
Comment (bamx23) [inline Sources/MapboxCoreNavigation/Router.swift:416]: `bestCandidate.element.route.description` or `target`?

File: Sources/MapboxCoreNavigation/Router.swift [hunk 4]
```diff
@@ -404,10 +404,18 @@ extension InternalRouter where Self: Router {
...
-            let rightDistance = right.element.description.minimumEditDistance(to: target)
-            return leftDistance < rightDistance
-        }?.offset
+        
+        guard let bestCandidate = map({
+            (route: $0, editDistance: $0.description.minimumEditDistance(to: target))
+        }).enumerated().min(by: { $0.element.editDistance < $1.element.editDistance }) else { return nil }
+
+
+        // If the most similar route is still more than 50% different from the original route,
```
Comment (bamx23) [inline Sources/MapboxCoreNavigation/Router.swift:413]: Why 50%?

Comment (1ec5) [reply Sources/MapboxCoreNavigation/Router.swift:413]: This number came from https://github.com/mapbox/mapbox-navigation-ios/issues/3597#issuecomment-982362158. There are only ever two road names in the route description per leg, so a 50% match is roughly one of the two names matching (modulo any differences in name length or minor things like “South”). We could probably lower the threshold without much problem, but this becomes a bugfarm, then the per-step approach in https://github.com/mapbox/mapbox-navigation-android/issues/3116#issuecomment-655710099 would be a decent plan B.

File: Sources/MapboxCoreNavigation/Router.swift [hunk 5]
```diff
@@ -404,10 +404,18 @@ extension InternalRouter where Self: Router {
...
+        // we fallback to the fastest route which index is 0.
+        guard bestCandidate.element.route.description.count > 0 else { return 0 }
+        let similarityScore = Double(bestCandidate.element.editDistance) / Double(bestCandidate.element.route.description.count)
+        guard similarityScore < 0.5 else { return 0 }
```
Comment (1ec5) [inline Sources/MapboxCoreNavigation/Router.swift:417]: Thanks, I inverted this logic in https://github.com/mapbox/mapbox-navigation-ios/issues/3597#issuecomment-982362158 because I mistook `score` for a measure of similarity, but in fact it’s still based on the edit distance, so it’s really a relative difference.

File: Sources/MapboxCoreNavigation/Router.swift (hunk 3 above)
Comment (1ec5) [reply Sources/MapboxCoreNavigation/Router.swift:416]: Good question, I don’t have a good intuition for this. Can we try out a few scenarios to see which would be more correct?

File: Sources/TestHelper/DummyRoute.swift (hunk 1 above)
Comment (1ec5) [reply Sources/TestHelper/DummyRoute.swift:14]: This is correct, since `Route.description` is based on `RouteLeg.name`:

https://github.com/mapbox/mapbox-directions-swift/blob/28a0bc98ca13ac7cbfa08d06f273dcd94503d696/Sources/MapboxDirections/DirectionsResult.swift#L191
https://github.com/mapbox/mapbox-directions-swift/blob/28a0bc98ca13ac7cbfa08d06f273dcd94503d696/Sources/MapboxDirections/RouteLeg.swift#L353

File: Tests/MapboxCoreNavigationTests/RerouteTests.swift [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+        let mostCommonIndex = candidates.index(mostSimilarTo: originalRoute)
+        XCTAssertEqual(mostCommonIndex, 2)
+    }
+
+    func testSelectsFastestRouteIfAllCandidatesAreVeryDifferent() {
+        let originalRoute = DummyRoute(description: "AAAAA")
+        let candidates = [
+            DummyRoute(description: "ABBBB"),
+            DummyRoute(description: "AABBB"),
```
Comment (1ec5) [inline Tests/MapboxCoreNavigationTests/RerouteTests.swift:24]: Let’s add tests where the descriptions are of different lengths.

File: Tests/MapboxCoreNavigationTests/RerouteTests.swift [hunk 7]
```diff
@@ -0,0 +1,34 @@
+import MapboxDirections
+@testable import MapboxCoreNavigation
+import XCTest
+import TestHelper
+
+class RerouteTests: XCTestCase {
+
+    func testSelectsMostSililarRouteForRerouting() {
```
Comment (1ec5) [inline Tests/MapboxCoreNavigationTests/RerouteTests.swift:8]:
```suggestion
    func testSelectsMostSimilarRouteForRerouting() {
```

**Review by 1ec5 — APPROVED** [review]
Suggested changelog entry:

* When rerouting the user, if none of the new routes is very similar to the original route selection, the `Router` now follows the most optimal route, not a route that is only marginally similar. ([#3664](https://github.com/mapbox/mapbox-navigation-ios/pull/3664))

File: Sources/MapboxCoreNavigation/Router.swift (hunk 3 above)
Comment (bamx23) [reply Sources/MapboxCoreNavigation/Router.swift:416]: If the best candidate is when we remove letters from the `target`, then we might get more than 1 here. I.e.
```
target = "AAAAAA"
bestCandidate = "AA"
editDistance = 4 // remove any 4 letters
score = 4 / len("AA") = 4 / 2 = 2
```
At the same time, if the best candidate is when you add letters, then we can get the same "larger than 1" if compared to the `target` length. 

We can divide by the sum of the sizes to guarantee that we get the value within `0...1`, but it's a bit more difficult to interpret such a score, i.e. 
```
"AAAAAA" -> "AAAAAA" = 0 / 12 = 0.00
"AAAAAA" -> "AAAAA"  = 1 / 11 = 0.09
"AAAAAA" -> "AAAA"   = 2 / 10 = 0.20
"AAAAAA" -> "AAA"    = 3 /  9 = 0.33
"AAAAAA" -> "AA"     = 4 /  8 = 0.50
"AAAAAA" -> "A"      = 5 /  7 = 0.71
"AAAAAA" -> ""       = 6 /  6 = 1.00

"AAAAAA" -> "AAAABB" = 2 / 12 = 0.17
"AAAAAA" -> "AAABBB" = 3 / 12 = 0.25
"AAAAAA" -> "AABBBB" = 4 / 12 = 0.33
"AAAAAA" -> "BBBBBB" = 6 / 12 = 0.50

"AAAAAA" -> "BBB"    = 6 /  9 = 0.6

"AAAAAA" -> "AAAAAAC"     = 1 / 13 = 0.08
"AAAAAA" -> "AAAAAACCCC"  = 4 / 16 = 0.25
"AAAAAA" -> "AAAAAACCCCC" = 5 / 17 = 0.29
```
Maybe we can compare the score to `0.25` as for "replacing the half of the string" or similar.

Comment (azarovalex) [reply Sources/MapboxCoreNavigation/Router.swift]: I added some tests for different candidate lengths and it seems like @bamx23's algorithm might show better results for some edge cases.
It's now a little bit harder to understand, but I added a comment with a link to this explanation.

Even if we are unable to select the most common route we will fallback to the fastest route which not as bad.

**Comment by 762caliber** [comment]
Hi, 


checking in for a update for NCS [#335](https://github.com/mapbox-collab/ncs-collab/issues/335).

Could we confirm that the PR have completed / fixed in which version of code please.


cc @bentruscello , @venkat-b-1983 , @OttyLab

**Comment by 762caliber** [comment]
Hi,

checking in for a update for NCS #335.

Could we confirm that the PR have completed / fixed in which version of code please.

cc @bentruscello , @venkat-b-1983 , @OttyLab
